### PR TITLE
#883: Wiremock: Fix for Spring context always being set to Dirty.

### PR DESCRIPTION
**Pull request by boothen**
This is a fix for #883.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@boothen Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=884)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=884) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@boothen Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=884)!

File: spring-cloud-contract-wiremock/src/test/java/org/springframework/cloud/contract/wiremock/AutoConfigureWireMockRandomPortApplicationTests.java [hunk 1]
```diff
@@ -38,11 +38,21 @@
...
+
 	@Test
 	public void contextLoads() throws Exception {
 		stubFor(get(urlEqualTo("/test")).willReturn(aResponse()
 				.withHeader("Content-Type", "text/plain").withBody("Hello World!")));
 		assertThat(this.service.go()).isEqualTo("Hello World!");
 	}
 
+	@Test
+	public void contextIsNotReload() throws Exception {
```
Comment (OlgaMaciaszek) [inline spring-cloud-contract-wiremock/src/test/java/org/springframework/cloud/contract/wiremock/AutoConfigureWireMockRandomPortApplicationTests.java]: Maybe change test method name to something like `dynamicPortsAreResolvedCorrectly`? You don't really assert here if the context is reloaded.

Comment (boothen) [reply spring-cloud-contract-wiremock/src/test/java/org/springframework/cloud/contract/wiremock/AutoConfigureWireMockRandomPortApplicationTests.java]: I've updated the code as directed. Thanks for coming back to me so quick.

Comment (OlgaMaciaszek) [reply spring-cloud-contract-wiremock/src/test/java/org/springframework/cloud/contract/wiremock/AutoConfigureWireMockRandomPortApplicationTests.java]: Sure. However, I see the build is failing exactly with the issue that checking ports was introduced to fix, so it looks like, there has been no context cleanup where it needed to happen.

Comment (boothen) [reply spring-cloud-contract-wiremock/src/test/java/org/springframework/cloud/contract/wiremock/AutoConfigureWireMockRandomPortApplicationTests.java]: Sorry about that. Late night code changes without applicable tests never workout well. I've fixed the logic around portIsFixed.

The tests were failing as previous classes weren't being marked to have the Spring context reloaded. The code worked before the change as it always reloaded the Spring context. I don't think it's a good idea that a test can pass in isolation but fail when run as part of a suite. Therefore, I've added a new test the applicable test classes to verify that the portIsFixed or not fixed depending on the configuration. I hope that's ok?

**Comment by OlgaMaciaszek** [comment]
@boothen Thanks for fixing the tests. I also think it's a good idea to have somewhere a test that verifies fixed ports and a test that verifies not fixed ports. But I don't think we should add more of these same tests to different classes, as they verify a single code execution path - the behaviour of ports does not depend on other conditions set in different tests, so please leave one of each (portFixed, porstNotFixed) and remove the rest. That should be the last change :), and then we can merge.

**Comment by boothen** [comment]
Removed most of the portIsFixed/portNotFixed tests. I've left in 4 which cover different port configurations. i.e port=0, port=9999, port = 0 & httpsPort=9999, port = 0 and httpsPort =0.

**Comment by marcingrzejszczak** [comment]
Hey @boothen have you rebase against master? Cause I see a couple of checkstyle issues that should have broken your build.

**Comment by boothen** [comment]
Sorry about that. I had the wrong codestyle configured. Should be ok now?

**Comment by OlgaMaciaszek** [comment]
Hello @boothen Some tests have started failing again. Have you tried merging against master and then rebuilding this locally? Does it pass locally?

**Comment by OlgaMaciaszek** [comment]
Ok, @boothen I see it's passing now. Merging. Thanks for the fix.

**Comment by boothen** [comment]
Hi @OlgaMaciaszek,

I believe there could be random test failures due to the running order of tests. The test classes with random ports won't dirty the Spring Context. This means if 2 test classes are run consecutively that use random ports, the 2nd class would use the same as the 1st. Even if there are differences in the test configuration.

For instance, if AutoConfigureWireMockStubsAndMultipleFilesApplicationTests was executed followed by AutoConfigureWireMockStubsApplicationTests, the context wouldn't be restarted to use the new stub mapping in the 2nd test class and so could fail.

The workaround for this issue would be to add the DirtiesContext annotation to all test classes that use the random ports.

Thanks.
Matt.

**Comment by marcingrzejszczak** [comment]
> The workaround for this issue would be to add the DirtiesContext annotation to all test classes that use the random ports.

This is the very essence of this PR. Not to do this. If a user does the same in their codebase, will we ask them to use `DirtiesContext` ?

**Comment by boothen** [comment]
The issue fixed is that the context was always being set to dirty whether dynamic or fixed ports were used. With the fix in place dynamic ports now won't reload the context.

However, some automated tests with dynamic ports use different wiremock configuration as part of their tests, so need the context reloading. e.g. files or stub configuration are different.

A user in their codebase would more likely be using the same wiremock configuration across their tests, so less likely to need to use the DirtiesContext.

**Pull request merged**